data/src/main/java/com/linkedin/data/codec/entitystream/ProtobufDataDecoder.java

```diff
@@ -0,0 +1,891 @@
+    {
+      Token refToken = readInt32();
+      if (refToken == NOT_AVAILABLE)
+      {
+        return NOT_AVAILABLE;
+      }
+      if ((_stringValue = _symbolTable.getSymbolName(_intValue)) == null)
+      {
+        throw new DataDecodingException("Error decoding string reference");
+      }
+      return STRING;
+    }
+
+    private Token finishToken(Token token)
+    {
+      _currentToken = token;
+      switch (_currentToken)
+      {
+        case START_OBJECT:
+          if (_currComplexObjTokenSize > 0)
+          {
+            _complexObjTokenSizeStack.push(_currComplexObjTokenSize);
+          }
+          _currComplexObjTokenSize = _intValue << 1;
+          break;
+        case START_ARRAY:
+          if (_currComplexObjTokenSize > 0)
+          {
+            _complexObjTokenSizeStack.push(_currComplexObjTokenSize);
+          }
+          _currComplexObjTokenSize = _intValue;
+          break;
+        case NOT_AVAILABLE:
+          break;
+        default:
+          _currComplexObjTokenSize--;
+      }
+      return _currentToken;
+    }
+
+    @Override
+    public int getComplexObjSize()
+    {
+      return _currentToken == START_OBJECT || _currentToken == START_ARRAY ? _intValue : -1;
+    }
+
+    @Override
+    public String getString() throws IOException
+    {
+      if (_currentToken != STRING)
+      {
+        throw new DataDecodingException("Unexpected call: String value is not available");
+      }
+      return _stringValue;
+    }
+
+    @Override
+    public ByteString getRawBytes() throws IOException {
+      if (_currentToken != RAW_BYTES)
+      {
+        throw new DataDecodingException("Unexpected call: Raw bytes value is not available");
+      }
+      return ByteString.unsafeWrap(_bytesValue);
+    }
+
+    @Override
+    public int getIntValue() throws IOException
+    {
+      if (_currentToken != INTEGER)
+      {
+        throw new DataDecodingException("Unexpected call: int value is not available");
+      }
+      return _intValue;
+    }
+
+    @Override
+    public long getLongValue() throws IOException
+    {
+      if (_currentToken != LONG)
+      {
+        throw new DataDecodingException("Unexpected call: Raw bytes value is not available");
+      }
+      return _longValue;
+    }
+
+    @Override
+    public float getFloatValue() throws IOException
+    {
+      if (_currentToken != FLOAT)
+      {
+        throw new DataDecodingException("Unexpected call: Raw bytes value is not available");
+      }
+      return Float.intBitsToFloat(_intValue);
+    }
+
+    @Override
+    public double getDoubleValue() throws IOException
+    {
+      if (_currentToken != DOUBLE)
+      {
+        throw new DataDecodingException("Unexpected call: Raw bytes value is not available");
+      }
+      return Double.longBitsToDouble(_longValue);
+    }
+
+    /*
+    * Non blocking ProtoReader Implementation
+    */
+
+    private Token readASCIIString() throws IOException {
```

Comment:
Add doc here, explain the different scenarios:
1. new string, fully readable
2. new string, partially readable
3. reading remaining part of a string.
and include details of the different variables used for each scenario. For example _intValue represents the remaining chars to read for scenario 3.

Reply:
added comments for 3 cases.
commented on _intValue variable, also using local variable as remainingSize to signify the same.